pennylane/transforms/commutation_dag.py

```diff
@@ -0,0 +1,1049 @@
+        )
+
+        target_control = intersection(operation1.wires, operation2.control_wires)
+        target_target = intersection(operation1.wires, target_wires_2)
+
+        # Case 4.1: control and target 2 overlap with target 1
+        if target_control and target_target:
+            return bool(commutation_map[operation1.name][position[control_base_2]]) and bool(
+                commutation_map[operation1.name][position[control_base_2]]
+            )
+
+        # Case 4.2: control operation 2 overlap with target 1
+        if target_control and not target_target:
+            return bool(commutation_map[operation1.name][position["ctrl"]])
+
+        # Case 4.3: target 1 overlaps with target 2
+        if not target_control and target_target:
+            return bool(commutation_map[operation1.name][position[control_base_2]])
+
+    # Case 5: no controlled operations
+    # Case 5.1: no controlled operations we simply check the commutation table
+    return bool(commutation_map[operation1.name][position[operation2.name]])
+
+
+def _merge_no_duplicates(*iterables):
+    """Merge K list without duplicate using python heapq ordered merging.
+
+    Args:
+        *iterables: A list of k sorted lists
+
+    Yields:
+        Iterator: List from the merging of the k ones (without duplicates)
+    """
+    last = object()
+    for val in heapq.merge(*iterables):
+        if val != last:
+            last = val
+            yield val
+
+
+class CommutationDAGNode:
+    r"""Class to store information about a quantum operation in a node of the
+    commutation DAG.
+
+    Args:
+        op (.Operation): PennyLane operation.
+        wires (.Wires): Wires on which the operation acts on.
+        node_id (int): ID of the node in the DAG.
+        successors (array[int]): List of the node's successors in the DAG.
+        predecessors (array[int]): List of the node's predecessors in the DAG.
+        reachable (bool): Attribute used to check reachability by pairwise commutation.
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-few-public-methods
+
+    __slots__ = [
+        "op",
+        "wires",
+        "target_wires",
+        "control_wires",
+        "node_id",
+        "successors",
+        "predecessors",
+        "reachable",
+    ]
+
+    def __init__(
+        self,
+        op=None,
+        wires=None,
+        target_wires=None,
+        control_wires=None,
+        successors=None,
+        predecessors=None,
+        reachable=None,
+        node_id=-1,
+    ):
+        self.op = op
+        self.wires = wires
+        self.target_wires = target_wires
+        self.control_wires = control_wires if control_wires is not None else []
+        self.node_id = node_id
+        self.successors = successors if successors is not None else []
+        self.predecessors = predecessors if predecessors is not None else []
+        self.reachable = reachable
```

Comment:
Minor comment, but it could be nice to document these attributes - they are currently showing up in the docs blank!

https://pennylane--1712.org.readthedocs.build/en/1712/code/api/pennylane.transforms.CommutationDAGNode.html

![image](https://user-images.githubusercontent.com/2959003/155729646-f335642b-307a-4152-8ad6-6d8b1ead1d5e.png)

Reply:
I will add all attributes to the documentation.